File: doric-android/doric/src/main/java/pub/doric/resource/DoricResourceManager.java
/*
 * Copyright [2021] [Doric.Pub]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pub.doric.resource;

import com.github.pengfeizhou.jscore.JSArrayBuffer;
import com.github.pengfeizhou.jscore.JSObject;
import com.github.pengfeizhou.jscore.JSValue;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import pub.doric.DoricContext;

/**
 * @Description: This manages all resource loaders
 * @Author: pengfei.zhou
 * @CreateDate: 2021/10/20
 */
public class DoricResourceManager {
    private final Map<String, DoricResourceLoader> mResourceLoaders = new HashMap<>();

    public synchronized void registerLoader(DoricResourceLoader loader) {
        mResourceLoaders.put(loader.resourceType(), loader);
    }

    public synchronized void unRegisterLoader(DoricResourceLoader loader) {
        mResourceLoaders.remove(loader.resourceType());
    }

    @Nullable
    @UiThread
    public synchronized DoricResource load(@NonNull DoricContext doricContext,
                                           @NonNull JSObject resource) {
        String resId = resource.getProperty("resId").asString().value();
        String type = resource.getProperty("type").asString().value();
        String identifier = resource.getProperty("identifier").asString().value();
        DoricResource doricResource = doricContext.getCachedResource(resId);
        if (doricResource == null) {
            DoricResourceLoader loader = mResourceLoaders.get(type);
            if (loader != null) {
                doricResource = loader.load(doricContext, identifier);
                if (doricResource instanceof DoricArrayBufferResource) {
                    JSArrayBuffer buffer = resource.getProperty("data").asArrayBuffer();
                    ((DoricArrayBufferResource) doricResource).setValue(buffer);
                } else if (doricResource instanceof DoricRemoteResource) {
                    JSValue headers = resource.getProperty("headers");
                    if (headers.isObject()) {
                        JSObject headersObject = headers.asObject();
                        for (Map.Entry<String, JSValue> entry : headersObject.value().entrySet()) {
                            ((DoricRemoteResource) doricResource).setHeader(entry.getKey(),
                                    entry.getValue().asString().value());
                        }
                    }
                }
                doricContext.cacheResource(resId, doricResource);
            }
        }
        return doricResource;
    }
}

File: doric-iOS/Pod/Classes/Resource/DoricResourceManager.m
/*
 * Copyright [2021] [Doric.Pub]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//
// Created by pengfei.zhou on 2021/10/22.
//

#import "DoricResourceManager.h"
#import "DoricContext.h"
#import "DoricArrayBufferResource.h"
#import "DoricRemoteResource.h"

@interface DoricResourceManager ()
@property(nonatomic, strong) NSMutableDictionary <NSString *, id <DoricResourceLoader>> *loaders;
@property(nonatomic, strong) dispatch_queue_t mapQueue;
@end

@implementation DoricResourceManager
- (instancetype)init {
    if (self = [super init]) {
        _loaders = [NSMutableDictionary new];
        _mapQueue = dispatch_queue_create("doric.resource", DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

- (void)registerLoader:(id <DoricResourceLoader>)loader {
    dispatch_sync(self.mapQueue, ^{
        self.loaders[loader.resourceType] = loader;
    });
}

- (void)unRegisterLoader:(id <DoricResourceLoader>)loader {
    dispatch_sync(self.mapQueue, ^{
        [self.loaders removeObjectForKey:loader.resourceType];
    });
}

- (__kindof DoricResource *)load:(NSDictionary *)resource
                     withContext:(DoricContext *)context {
    NSString *type = resource[@"type"];
    NSString *identifier = resource[@"identifier"];
    NSString *resId = resource[@"resId"];
    __block __kindof DoricResource *doricResource;
    dispatch_sync(self.mapQueue, ^() {
        doricResource = [context.cachedResources objectForKey:resId];
        if (!doricResource) {
            id <DoricResourceLoader> loader = self.loaders[type];
            doricResource = [loader load:identifier withContext:context];
            if ([doricResource isKindOfClass:DoricArrayBufferResource.class]) {
                ((DoricArrayBufferResource *) doricResource).data = resource[@"data"];
            } else if ([doricResource isKindOfClass:DoricRemoteResource.class]) {
                NSDictionary *headers = resource[@"headers"];
                if (headers) {
                    [headers enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop) {
                        [(DoricRemoteResource *) doricResource setHeaderWithKey:key
                                                                      withValue:value];
                    }];
                }
            }
            [context.cachedResources setObject:doricResource forKey:resId];
        }
    });
    return doricResource;
}

@end
